### PR TITLE
VZ-3827 Add ability to set a DestinationRule with HTTP Cookie for session affinity

### PR DESCRIPTION
**Pull request by bryantidd**
# Description

Please include a summary of the change and which issue is fixed.  If there are any dependencies, for example on a PR in another repository, please list those.

Fixes VZ-9999

# Checklist 

As the author of this PR, I have:

- [x] Checked that I included or updated copyright and license notices in all files that I altered
- [ ] Added or updated unit tests for any new functions I added
- [ ] Added or updated integration tests if appropriate
- [x] Added or updated acceptance tests if appropriate

Code reviewer, please confirm this PR:

- [ ] Addressed the requirement and meets the acceptance criteria
- [ ] Does not introduce unrelated or spurious changes
- [ ] Does not introduce any unapproved dependency
- [ ] Makes sense and it easy to understand, and/or difficult areas of code are clearly documented so that they can be understood

### CODE REVIEW

File: application-operator/controllers/ingresstrait/ingresstrait_controller.go [hunk 1]
```diff
@@ -534,6 +539,56 @@ func (r *Reconciler) mutateVirtualService(virtualService *istioclient.VirtualSer
...
+										Ttl:  ptypes.DurationProto(rule.Destination.HTTPCookie.TTL)},
+								},
+							},
+						},
+					},
+				},
+			},
+		}
+
+		res, err := controllerutil.CreateOrUpdate(ctx, r.Client, destinationRule, func() error {
```
Comment (desagar) [inline application-operator/controllers/ingresstrait/ingresstrait_controller.go:571]: controllerutil.CreateOrUpdate first does a GET of the resource _into_ the object corresponding to the pointer you pass in, and _then_ calls the mutate function to give you a chance to edit it. 

The mutate function is where you're supposed to take the resource (if any) that already exists, and modify it.

If the destination rule already exists and is not in sync with the ingress trait, the way you're doing might not correct it to be in sync. The typical way to achieve this is:

- Create a destinationRule object that only has namespace and name populated
- Call controllerutil.CreateOrUpdate
- Do _all_ the manipulation inside the mutate function

Comment (bryantidd) [reply application-operator/controllers/ingresstrait/ingresstrait_controller.go:571]: Will do, Thanks

Comment (desagar) [reply application-operator/controllers/ingresstrait/ingresstrait_controller.go:554]: and given how complicated this is, probably would be good to add some coverage in the unit test

**Review by wmhopkins — APPROVED** [review]
lgtm.

**Pull request merged**